packages/waku/src/router/create-pages.ts

```diff
@@ -457,3 +458,277 @@ export function createPages<
+      const paths: {
+        pattern: string;
+        path: PathSpec;
+        components: Record<string, { isStatic: boolean }>;
+        noSsr: boolean;
+      }[] = [];
+      for (const [path, pathSpec] of fixedPathSet) {
+        const noSsr = noSsrSet.has(pathSpec);
+        const isStatic = (() => {
+          for (const [_, [layoutPathSpec]] of dynamicLayoutPathMap) {
+            if (hasPathSpecPrefix(layoutPathSpec, pathSpec)) {
+              return false;
+            }
+          }
+          return true;
+        })();
+
+        paths.push({
+          pattern: path2regexp(parsePathWithSlug(path)),
+          path: pathSpec,
+          components: { 'layout:/': { isStatic } },
+          noSsr,
+        });
+      }
+      for (const [path, [pathSpec]] of dynamicPagePathMap) {
+        const noSsr = noSsrSet.has(pathSpec);
+        paths.push({
+          pattern: path2regexp(parsePathWithSlug(path)),
+          path: pathSpec,
+          components: { 'page:/': { isStatic: false } },
+          noSsr,
+        });
+      }
+      for (const [path, [pathSpec]] of wildcardPagePathMap) {
+        const noSsr = noSsrSet.has(pathSpec);
+        paths.push({
+          pattern: path2regexp(parsePathWithSlug(path)),
+          path: pathSpec,
+          components: { 'page:/': { isStatic: false } },
+          noSsr,
+        });
+      }
+      return paths;
+    },
+    renderRoute: async (path, options) => {
+      await configure();
+
+      const processSkip = <T>(elements: Record<string, T>) => {
+        if (!options.skip) {
+          return elements;
+        }
+        return Object.fromEntries(
+          Object.entries(elements).filter(([k]) => !options.skip!.includes(k)),
+        );
+      };
+
+      // TODO it would be good if routeMap can exist as a lookup for all routes in createPages
+
+      const result: Record<string, ReactNode> = {
+        root: rootItem
+          ? rootItem.component({ children: Children() }) // Is this right?
+          : DefaultRoot({ children: Children() }),
+      };
+
+      // TODO use pathSpec to trace through routes for nested layouts
+      // const pathSpec = parsePathWithSlug(path);
+
+      const layout =
+        dynamicLayoutPathMap.get('/')?.[1] ??
+        staticComponentMap.get(joinPath('/', 'layout'));
+      if (layout) {
+        result['layout:/'] = layout({ children: Children() });
+      }
+      if (path === '/') {
+        const pageComponent = (staticComponentMap.get('/') ??
+          dynamicPagePathMap.get('/')?.[1])!;
+        result['page:/'] = pageComponent({ children: Children() });
+        const routeChildren = [
+          (layout ? createElement(Slot, { id: 'layout:/' }, []) : null)!,
+          createElement(Slot, { id: 'page:/' }, []),
+        ].filter(Boolean);
+        result['route:/'] = createElement(
+          Slot,
+          { id: 'root' },
+          ...routeChildren,
+        );
+      }
+
+      return processSkip(result);
+    },
```

Comment:
This is the bulk of what's new so far. Does this look on the right track?

I still have a lot of testing and other handling to do, but I think some initial feedback on this draft would be helpful, thanks!

Reply:
It may not work as is, but I think it feels on the right track.